Sources/Commands/Workspace.swift

```diff
@@ -400,6 +412,20 @@ public class Workspace {
         try saveState()
     }
 
+    /// Creates a clone at the provided version and pins the repository at that version.
+    ///
+    /// - Parameters:
+    ///   - repository: The repository to pin. It should not be already pinned.
+    ///   - at: The version to pin at.
+    ///   - reason: The optional reason for pinning.
+    /// - Throws: WorkspaceOperationError, PinOperationError
+    func pin(repository: RepositorySpecifier, at version: Version, reason: String? = nil) throws {
+        // Clone the version we want to pin.
+        _ = try clone(specifier: repository, version: version)
```

Comment:
Why do we need to clone here?

Shouldn't we attempt pin, then normalize state? Also, shouldn't this attempt to resolve the pin before accepting it (in case it is invalid, better to say that here rather than lose previous pin state).